### PR TITLE
Invoke-WhiskeyExec task now supports ranges ( .., >=, <=, >, < ) for …

### PR DESCRIPTION
**Pull request by jeffprice6**
…`SuccessExitCode` config property

### CODE REVIEW

File: Whiskey/Tasks/Invoke-WhiskeyExec.ps1 [hunk 1]
```diff
@@ -120,19 +116,87 @@ function Invoke-WhiskeyExec
...
 
-    $successExitCode = 0
+    [int[]]$successExitCode = $null
+    $successExitRangeGreaterThan = $null
+    $successExitRangeLessThan = $null
+
     if ( $TaskParameter['SuccessExitCode'] )
     {
-        $successExitCode = $TaskParameter['SuccessExitCode']
+        $TaskParameter['SuccessExitCode'] | ForEach-Object {
```
Comment (splatteredbits) [inline Whiskey/Tasks/Invoke-WhiskeyExec.ps1]: Move this whole block to after the process runs. Evaluate the exit code after the process exits. If you move it there, the logic for evaluating can be simpler. As soon as you find a success exit code matches the exit code, you can return from the function. 

Here's the code I came up with:

```
$exitCode = $process.ExitCode

$successExitCodes = $TaskParameter['SuccessExitCode']
if( -not $successExitCodes )
{
    $successExitCodes = '0'
}

foreach( $successExitCode in $successExitCodes )
{
    if( $successExitCode -match '^(\d)+$' )
    {
        if( $exitCode -eq [int]$Matches[1] )
        {
            Write-Verbose -Message ('  Exit code {0} = {1}' -f $exitCode,$Matches[1])
            return
        }
    }

    if( $successExitCode -match '^(<|<=|>=|>)\s*(\d)+$' )
    {
        $operator = $Matches[1]
        $successExitCode = [int]$Matches[2]
        switch( $operator )
        {
            '<'
            {
                if( $exitCode -lt $successExitCode )
                {
                    Write-Verbose -Message ('  Exit code {0} < {1}' -f $exitCode,$successExitCode)
                    return
                }
            }
            '<=' 
            {
                if( $exitCode -le $successExitCode )
                {
                    Write-Verbose -Message ('  Exit code {0} <= {1}' -f $exitCode,$successExitCode)
                    return
                }
            }
            '>='
            {
                if( $exitCode -ge $successExitCode )
                {
                    Write-Verbose -Message ('  Exit code {0} >= {1}' -f $exitCode,$successExitCode)
                    return
                }
            }
            '>' 
            {
                if( $exitCode -gt $successExitCode )
                {
                    Write-Verbose -Message ('  Exit code {0} > {1}' -f $exitCode,$successExitCode)
                    return
                }
            }
        }
    }

    if( $successExitCode -match '^(\d)+\.\.\(d+)$' )
    {
        if( [int]$Matches[1] -le $exitCode -and $exitCode -le [int]$Matches[2] )
        {
            Write-Verbose -Message ('  Exit code {0} <= {1} <= {2}' -f $Matches[1],$exitCode,$Matches[2])
            return
        }
    }
}
```

File: Whiskey/Tasks/Invoke-WhiskeyExec.ps1 [hunk 2]
```diff
@@ -120,19 +116,87 @@ function Invoke-WhiskeyExec
...
+    }
+    
+    if ( !$successExitRangeLessThan )
+    {
+        $successExitRangeLessThan = $exitCode - 1
+    }
+
+    if ( !($exitCode -in $successExitCode) -and !($exitCode -gt $successExitRangeGreaterThan) -and !($exitCode -lt $successExitRangeLessThan) )
     {
         Stop-WhiskeyTask -TaskContext $TaskContext -Message ('''{0}'' returned with an exit code of ''{1}'', which is not one of the expected ''SuccessExitCode'' of ''{2}''. View the build output to see why the executable''s process failed.' -F $TaskParameter['Path'],$exitCode,$successExitCode -join ',')
```
Comment (splatteredbits) [inline Whiskey/Tasks/Invoke-WhiskeyExec.ps1]: Change this to not mention the success exit codes.

File: Test/Invoke-WhiskeyExec.Tests.ps1 [hunk 3]
```diff
@@ -269,6 +269,17 @@ Describe 'Invoke-WhiskeyExec.when given success exit codes' {
     ThenTaskSuccess
 }
 
+Describe 'Invoke-WhiskeyExec.when given success exit codes with ''='' prefix' {
+    Init
+    GivenExecutableFile 'executable.bat' 'exit 123'
+    GivenPath 'executable.bat'
+    GivenSuccessExitCode '=123'
```
[the remark this answers is not in the document]
Comment (jeffprice6) [reply Test/Invoke-WhiskeyExec.Tests.ps1]: Removed

File: Test/Invoke-WhiskeyExec.Tests.ps1 [hunk 4]
```diff
@@ -280,6 +291,116 @@ Describe 'Invoke-WhiskeyExec.when executable exits with non-success exit code' {
     ThenTaskFailedWithMessage 'not one of the expected ''SuccessExitCode'''
 }
 
+Describe 'Invoke-WhiskeyExec.when given a range ''..'' of success exit codes' {
+    Init
+    GivenExecutableFile 'executable.bat' 'exit 123'
+    GivenPath 'executable.bat'
+    GivenSuccessExitCode 120..130
```
[the remark this answers is not in the document]
Comment (jeffprice6) [reply Test/Invoke-WhiskeyExec.Tests.ps1]: Updated

File: Test/Invoke-WhiskeyExec.Tests.ps1 [hunk 5]
```diff
@@ -280,6 +291,116 @@ Describe 'Invoke-WhiskeyExec.when executable exits with non-success exit code' {
...
+    ThenExecutableRan
+    ThenSpecifiedArgumentsWerePassed
+    ThenTaskSuccess
+}
+
+Describe 'Invoke-WhiskeyExec.when given a range ''..'' and exits with code outside success range' {
+    Init
+    GivenExecutableFile 'executable.bat' 'exit 133'
+    GivenPath 'executable.bat'
+    GivenSuccessExitCode 120..130
```
[the remark this answers is not in the document]
Comment (jeffprice6) [reply Test/Invoke-WhiskeyExec.Tests.ps1]: Updated

File: Whiskey/Tasks/Invoke-WhiskeyExec.ps1 [hunk 6]
```diff
@@ -120,19 +116,87 @@ function Invoke-WhiskeyExec
...
     {
-        $successExitCode = $TaskParameter['SuccessExitCode']
+        $TaskParameter['SuccessExitCode'] | ForEach-Object {
+            
+            if( $_ -match '^\d+$' )
+            {
+                $successExitCode += $_
+            }
+            
+            elseif( $_ -match '^\=\s*\d+$' )
```
[the remark this answers is not in the document]
Comment (jeffprice6) [reply Whiskey/Tasks/Invoke-WhiskeyExec.ps1]: Agreed, removed.

File: Whiskey/Tasks/Invoke-WhiskeyExec.ps1 (hunk 2 above)
Comment (jeffprice6) [reply Whiskey/Tasks/Invoke-WhiskeyExec.ps1]: Updated

**Pull request merged**